src/background.js

```diff
@@ -342,20 +354,66 @@ const rollout = {
     // Determine to skip additional heuristics (by presence of an enterprise policy)
     if (policyEnableDoH === "no_policy_set") {
       // Resetting skipHeuristicsCheck in case a user had a policy and then removed it!
-      await rollout.setSetting("doh-rollout.skipHeuristicsCheck", false);
+      await rollout.setSetting(DOH_SKIP_HEURISTICS_PREF, false);
     } else {
       // Don't check for prefHasUserValue if policy is set to disable DoH
-      await rollout.setSetting("doh-rollout.skipHeuristicsCheck", true);
+      await rollout.setSetting(DOH_SKIP_HEURISTICS_PREF, true);
     }
     return;
 
   },
 
+  async migrateLocalStoragePrefs() {
+
+    // Migrate updated local storage item names. If this has already been done once, skip the migration
+    const isMigrated = await browser.experiments.preferences.getBoolPref(DOH_BALROG_MIGRATION_PREF, false);
+
+    if (isMigrated) {
+      log("User has been migrated.");
+      return;
+    }
+
+    // Check all local storage keys from v1.0.4 users and migrate them to prefs.
+    // This only applies to keys that have a value.
+    const legacyLocalStorageKeys = [
+      "doneFirstRun",
+      "skipHeuristicsCheck",
+      DOH_ENABLED_PREF,
+      TRR_MODE_PREF,
+      DOH_SELF_ENABLED_PREF,
+      DOH_PREVIOUS_TRR_MODE_PREF,
+      DOH_DOORHANGER_SHOWN_PREF,
+      DOH_DOORHANGER_USER_DECISION_PREF,
+      DOH_DISABLED_PREF,
+      DOH_SKIP_HEURISTICS_PREF,
+      DOH_DONE_FIRST_RUN_PREF
+    ];
+
+    for (let item of legacyLocalStorageKeys) {
+      let data = await browser.storage.local.get(item);
+      let value = data[item];
+
+      if (data.hasOwnProperty(item)) {
+        let migratedName = item;
+
+        if (!item.startsWith("doh-rollout.")){
+          migratedName = "doh-rollout." + item;
+        }
+
+        await this.setSetting(migratedName, value);
```

Comment:
So, just wanted to point something out: `setSetting` is inferring the type of pref to set (char/bool/int) based on the type of the value. There's a class of bugs that result from this when the pref already exists but has a default value of a different type, in which case setFooPref will throw an error. This is probably a non-issue in this code since these prefs don't/shouldn't have default values, but just something to watch out for.

Reply:
Noted!